File: src/main/java/com/almasb/zeph/ui/CharInfoView.java
package com.almasb.zeph.ui;

import com.almasb.fxgl.dsl.FXGL;
import com.almasb.fxgl.ui.MDIWindow;
import com.almasb.zeph.character.CharacterEntity;
import com.almasb.zeph.combat.Attribute;
import com.almasb.zeph.combat.Stat;
import javafx.beans.binding.Bindings;
import javafx.beans.binding.StringBinding;
import javafx.beans.property.SimpleStringProperty;
import javafx.geometry.Orientation;
import javafx.scene.Cursor;
import javafx.scene.ImageCursor;
import javafx.scene.control.Separator;
import javafx.scene.control.Tooltip;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Pane;
import javafx.scene.layout.VBox;
import javafx.scene.paint.Color;
import javafx.scene.text.Font;
import javafx.scene.text.Text;

public class CharInfoView extends MDIWindow {

    public CharInfoView(CharacterEntity player) {
        //super("Char Info", WindowDecor.MINIMIZE);

        relocate(0, 240);

        //setBackgroundColor(Color.rgb(25, 25, 133, 0.4));
        setPrefSize(340, 340);
        setCanResize(false);

        Font font = Font.font("Lucida Console", 14);
        Font largerFont = Font.font("Lucida Console", 16);
        Cursor cursorQuestion = new ImageCursor(FXGL.image("ui/cursors/question.png"), 52, 10);

        VBox attrBox = new VBox(5);
        attrBox.setTranslateY(10);
        for (Attribute attr : Attribute.values()) {
            Text text = new Text();
            text.setFont(font);
            text.setFill(Color.WHITE);
            text.setCursor(cursorQuestion);
            text.textProperty().bind(player.getCharacterComponent().baseProperty(attr).asString(attr.toString() + ": %-3d"));

            Text tooltipText = new Text(attr.getDescription());
            tooltipText.setFill(Color.WHITE);
            tooltipText.setFont(largerFont);
            tooltipText.setWrappingWidth(200);

            Tooltip tooltip = new Tooltip();
            tooltip.setGraphic(tooltipText);

            Tooltip.install(text, tooltip);

            Text bText = new Text();
            bText.setFont(font);
            bText.setFill(Color.YELLOW);
            bText.visibleProperty().bind(player.getCharacterComponent().bonusProperty(attr).greaterThan(0));
            bText.textProperty().bind(player.getCharacterComponent().bonusProperty(attr).asString("+%d ")
                    .concat(player.getCharacterComponent().totalProperty(attr).asString("(%d)")));

            Text btn = new Text("+");
            btn.setCursor(Cursor.HAND);
            btn.setStroke(Color.YELLOWGREEN.brighter());
            btn.setStrokeWidth(3);
            btn.setFont(font);
            btn.visibleProperty().bind(player.getPlayerComponent().getAttributePoints().greaterThan(0)
                    .and(player.getCharacterComponent().baseProperty(attr).lessThan(100)));

            btn.setOnMouseClicked(event -> {
                player.getPlayerComponent().increaseAttribute(attr);
            });

            Pane box = new Pane();
            box.setPrefSize(160, 15);
            box.getChildren().addAll(text, bText, btn);

            bText.setTranslateX(70);
            btn.setTranslateX(155);

            attrBox.getChildren().add(box);
        }

        Text info = new Text();
        info.setFont(font);
        info.setFill(Color.WHITE);
        info.visibleProperty().bind(player.getPlayerComponent().getAttributePoints().greaterThan(0));
        info.textProperty().bind(new SimpleStringProperty("Attribute Points: ").concat(player.getPlayerComponent().getAttributePoints())
            .concat("\nSkill Points: ").concat(player.getPlayerComponent().getSkillPoints()));

        attrBox.getChildren().addAll(new Separator(), info);

        VBox statBox = new VBox(5);
        for (Stat stat : Stat.values()) {
            Text text = new Text();
            text.setFont(font);
            text.setFill(Color.WHITE);
            text.setCursor(cursorQuestion);
            text.textProperty().bind(player.getCharacterComponent().baseProperty(stat).asString(stat.toString() + ": %d"));

            Text tooltipText = new Text(stat.getDescription());
            tooltipText.setFill(Color.WHITE);
            tooltipText.setFont(largerFont);
            tooltipText.setWrappingWidth(200);

            Tooltip tooltip = new Tooltip();
            tooltip.setGraphic(tooltipText);

            Tooltip.install(text, tooltip);

            Text bText = new Text();
            bText.setFont(font);
            bText.setFill(Color.YELLOW);

            StringBinding textBinding = Bindings.when(player.getCharacterComponent().bonusProperty(stat).greaterThan(0))
                .then(player.getCharacterComponent().bonusProperty(stat).asString("+%d ")
                        .concat(player.getCharacterComponent().baseProperty(stat).add(player.getCharacterComponent().bonusProperty(stat)).asString("(%d)"))
                        .concat(stat.getMeasureUnit()))
                .otherwise(stat.getMeasureUnit());

            bText.textProperty().bind(textBinding);

            statBox.getChildren().add(new HBox(5, text, bText));
        }

        Pane root = new Pane(new HBox(10, attrBox, new Separator(Orientation.VERTICAL), statBox));

        setContentPane(root);

//        EventHandler<ActionEvent> handler = getRightIcons().get(0).getOnAction();
//        getRightIcons().get(0).setOnAction(e -> {
//            ScaleTransition st = new ScaleTransition(Duration.seconds(0.2), root);
//            st.setFromY(isMinimized() ? 0 : 1);
//            st.setToY(isMinimized() ? 1 : 0);
//            st.play();
//
//            handler.handle(e);
//        });
    }
}

File: src/main/kotlin/com/almasb/zeph/ui/CharInfoView.kt
package com.almasb.zeph.ui

import com.almasb.fxgl.dsl.FXGL
import com.almasb.zeph.character.CharacterEntity
import com.almasb.zeph.combat.Attribute
import com.almasb.zeph.combat.Stat
import javafx.beans.binding.Bindings
import javafx.beans.property.SimpleStringProperty
import javafx.geometry.Orientation
import javafx.scene.Cursor
import javafx.scene.ImageCursor
import javafx.scene.control.Separator
import javafx.scene.layout.*
import javafx.scene.paint.Color
import javafx.scene.text.Font
import javafx.scene.text.Text

/**
 *
 * @author Almas Baimagambetov (almaslvl@gmail.com)
 */
class CharInfoView(player: CharacterEntity) : Region() {

    init {
        background = Background(BackgroundFill(Color.BLACK, null, null))

        val font = Font.font("Lucida Console", 14.0)
        val cursorQuestion: Cursor = ImageCursor(FXGL.image("ui/cursors/question.png"), 52.0, 10.0)
        val attrBox = VBox(5.0)
        attrBox.translateY = 10.0

        for (attr in Attribute.values()) {
            val text = Text()
            text.font = font
            text.fill = Color.WHITE
            text.cursor = cursorQuestion
            text.textProperty().bind(player.characterComponent.baseProperty(attr).asString("$attr: %-3d"))

            val bText = Text()
            bText.font = font
            bText.fill = Color.YELLOW
            bText.visibleProperty().bind(player.characterComponent.bonusProperty(attr).greaterThan(0))
            bText.textProperty().bind(player.characterComponent.bonusProperty(attr).asString("+%d ")
                    .concat(player.characterComponent.totalProperty(attr).asString("(%d)")))

            val btn = Text("+")
            btn.cursor = Cursor.HAND
            btn.stroke = Color.YELLOWGREEN.brighter()
            btn.strokeWidth = 3.0
            btn.font = font
            btn.visibleProperty().bind(player.playerComponent!!.attributePoints.greaterThan(0)
                    .and(player.characterComponent.baseProperty(attr).lessThan(100)))
            btn.setOnMouseClicked {
                player.playerComponent!!.increaseAttribute(attr)
            }

            val box = Pane()
            box.setPrefSize(160.0, 15.0)
            box.children.addAll(text, bText, btn)
            bText.translateX = 70.0
            btn.translateX = 155.0

            box.setOnTooltipHover {
                it.setText(attr.description)
            }

            attrBox.children.add(box)
        }

        val info = Text()
        info.font = font
        info.fill = Color.WHITE
        info.visibleProperty().bind(player.playerComponent!!.attributePoints.greaterThan(0))
        info.textProperty().bind(SimpleStringProperty("Attribute Points: ").concat(player.playerComponent!!.attributePoints)
                .concat("\nSkill Points: ").concat(player.playerComponent!!.skillPoints))
        attrBox.children.addAll(Separator(), info)

        val statBox = VBox(5.0)
        for (stat in Stat.values()) {
            val text = Text()
            text.font = font
            text.fill = Color.WHITE
            text.cursor = cursorQuestion

            val statName = stat.toString().replace("_", " ")

            text.textProperty().bind(player.characterComponent.baseProperty(stat).asString("$statName: %d"))

            val bText = Text()
            bText.font = font
            bText.fill = Color.YELLOW
            val textBinding = Bindings.`when`(player.characterComponent.bonusProperty(stat).isNotEqualTo(0))
                    .then(player.characterComponent.bonusProperty(stat).asString("+%d ")
                            .concat(player.characterComponent.baseProperty(stat).add(player.characterComponent.bonusProperty(stat)).asString("(%d)"))
                            .concat(stat.measureUnit))
                    .otherwise(stat.measureUnit)

            bText.textProperty().bind(textBinding)

            val box = HBox(5.0, text, bText)
            box.setOnTooltipHover {
                it.setText(stat.description)
            }

            statBox.children.add(box)
        }

        children.add(HBox(10.0, attrBox, Separator(Orientation.VERTICAL), statBox))
    }
}